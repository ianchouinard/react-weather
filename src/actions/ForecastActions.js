
export const setZipCode = (payload) => {
  return {type: 'SET_ZIP', payload}
}

export const addZipHistory = (payload) => {
  return {type: 'ADD_ZIP_HISTORY', payload}
}

export const removeZipHistory = (payload) => {
  return {type: 'REMOVE_ZIP_HISTORY', payload}
}

export const setZipHistory = (payload) => {
  return {type: 'SET_ZIP_HISTORY', payload}
}

export const setUnitType = (payload) => {
  return {type: 'SET_UNIT_TYPE', payload}
}
