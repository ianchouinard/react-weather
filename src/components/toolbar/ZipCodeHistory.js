import React, { Component } from 'react'
import { MenuToggle } from '../MenuToggle'

export class ZipCodeHistory extends Component {

  state = {
    menuOpen: false
  }

  zipSelected = (e, zip) => {
    e.preventDefault()
    this.props.zipEntered(zip)
  }

  toggleMenu = () => {
    const menuState = this.state.menuOpen

    this.setState({
      menuOpen: !menuState
    })
  }

  render() {
    return (
      <nav className="zipCodeHistory">
        <MenuToggle open={this.state.menuOpen} toggleClick={this.toggleMenu} />
        <span>Recent Locations</span>
        <ul className={(this.state.menuOpen ? 'mobileOpen' : '')}>
          {this.props.zipCodeHistory.map((zipCode, index) => {
            return (
              <li key={ index }>
                <a
                  href="/"
                  title={`View weather for ${zipCode}`}
                  onClick={(e) => {this.zipSelected(e, zipCode)}}>{ zipCode }</a>
              </li>
            )
          })}
        </ul>
      </nav>
    )
  }

}
