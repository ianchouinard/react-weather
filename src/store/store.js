import { createStore, combineReducers } from 'redux'
import { ForecastMetaReducer } from './../reducers/ForecastMetaReducer'

const rootReducer = combineReducers({
  meta: ForecastMetaReducer
})

export const store = createStore(rootReducer)
