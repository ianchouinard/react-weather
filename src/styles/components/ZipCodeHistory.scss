
.zipCodeHistory {
  position: relative;
  display: flex;
  flex-wrap: nowrap;
  padding-bottom: 5px;
  height: 38px;
  overflow-y: hidden;
  z-index: 2;
  color: #fff;
  border-bottom: 1px solid #fff;

  span {
    margin-right: 10px;
    font-weight: 700;
    font-size: 1.2rem;
    white-space: nowrap;
  }

  ul {
    display: flex;
    flex-wrap: nowrap;
    vertical-align: middle;
    margin: 0;
    padding-left: 10px;
    min-height: 58px;
    overflow-x: auto;
    list-style: none;

    li {
      margin-right: 7px;

      a {
        display: inline-block;
        padding: 5px 10px;
        font-size: .9rem;
        color: #fff;
        border-radius: 3px;

        &:hover {
          background: rgba(250,250,250,.15);
          text-decoration: none;
        }
      }
    }
  }

  @media screen and (max-width: 500px) {
    display: block;
    padding: 5px 10px;
    height: 68px;
    border-bottom: none;

    ul {
      padding-left: 0;
    }
  }
}
