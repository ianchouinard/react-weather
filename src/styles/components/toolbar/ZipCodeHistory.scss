
.zipCodeHistory {
  position: relative;
  display: flex;
  flex-wrap: nowrap;
  padding-bottom: 5px;
  height: 38px;
  overflow-y: hidden;
  z-index: 2;
  color: #fff;
  border-bottom: 1px solid #fff;

  .menuToggle {
    display: none;
    margin-right: 10px;

    &+span {
      vertical-align: middle;
    }
  }

  span {
    margin-right: 10px;
    font-weight: 700;
    font-size: 1.2rem;
    white-space: nowrap;
  }

  ul {
    display: flex;
    flex-wrap: nowrap;
    vertical-align: middle;
    margin: 0;
    padding-left: 10px;
    min-height: 58px;
    overflow-x: auto;
    list-style: none;

    li {
      margin-right: 7px;

      a {
        display: inline-block;
        padding: 5px 10px;
        font-size: .9rem;
        color: #fff;
        border-radius: 3px;

        &:hover {
          background: rgba(250,250,250,.15);
          text-decoration: none;
        }
      }
    }
  }

  @media screen and (max-width: $mobile_small) {
    display: block;
    padding: 15px 0 5px 0;
    height: auto;
    border-bottom: none;
    background: $color_grey;

    .menuToggle {
      display: inline-block;
      margin-left: 15px;
    }

    ul {
      max-height: 0;
      min-height: 0;
      overflow: hidden;
      padding: 0 15px;
      margin-top: 10px;
      display: block;
      transition: all 350ms;

      &.mobileOpen {
        max-height: 200px;
        overflow-y: auto;
      }

      li {
        display: block;
        width: 100%;
        margin: 0;
      }
    }
  }
}
