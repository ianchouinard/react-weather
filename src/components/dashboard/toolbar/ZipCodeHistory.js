import React, { Component } from 'react'
import { MenuToggle } from '../../MenuToggle'
import { store } from '../../../store/store';
import { setZipCode } from '../../../actions/ForecastActions';

export class ZipCodeHistory extends Component {

  state = {
    menuOpen: false,
    zipCodeHistory: store.getState().zipCodeHistory
  }

  componentDidMount() {
    store.subscribe(() => {
      this.setState({
        zipCodeHistory: store.getState().zipCodeHistory
      })
    })
  }

  /**
   * User selected a zip code. 
   * Emit the zip code to the parent component
   */
  zipSelected = (e, zip) => {
    e.preventDefault()
    store.dispatch(setZipCode(zip))
  }

  /**
   * Toggle the mobile menu state
   */
  toggleMenu = () => {
    const menuState = this.state.menuOpen

    this.setState({
      menuOpen: !menuState
    })
  }

  render() {
    return (
      <nav className="zipCodeHistory">
        <MenuToggle open={this.state.menuOpen} toggleClick={this.toggleMenu} />
        <span>Recent Locations</span>
        <ul className={(this.state.menuOpen ? 'mobileOpen' : '')}>
          {this.state.zipCodeHistory.map((zipCode, index) => {
            return (
              <li key={ index }>
                <a
                  href="/"
                  title={`View weather for ${zipCode}`}
                  onClick={(e) => {this.zipSelected(e, zipCode)}}>{ zipCode }</a>
              </li>
            )
          })}
        </ul>
      </nav>
    )
  }

}
