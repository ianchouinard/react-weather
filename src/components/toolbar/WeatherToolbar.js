import React, {Component} from 'react'
import { ZipCodeHistory } from './ZipCodeHistory'
import background from './../../img/header_bg.jpg'
import backgroundRain from './../../img/header_bg_rain.jpg'
import backgroundSnow from './../../img/header_bg_snow.jpg'

export class WeatherToolbar extends Component {

  state = {
    zipCode: '03063',
    zipLooksValid: true
  }

  componentWillReceiveProps(props) {
    this.setState({
      zipCode: props.zipCode
    })
  }

  /**
   * Update the zipcode state.
   */
  zipChanged = (e) => this.setState({zipCode: e.target.value})

  /**
   * Emit the entered zip code.
   */
  zipEntered = (e) => {
    e.preventDefault()
    this.props.zipEntered(this.state.zipCode)
  }

  /**
   * Attempt to determine if the zip looks valid.
   * Will be disabled if true
   */
  zipLooksInvalid = () => {
    if (isNaN(parseInt(this.state.zipCode))) {
      return true
    }

    if (this.state.zipCode.length !== 5) {
      return true
    }

    return false
  }

  /**
   * Get header background based on weather condtions
   * weather code info here 
   * https://openweathermap.org/weather-conditions
   */
  getImageSource = () => {
    if (!this.props.conditions || !this.props.conditions.weather) {
      return ''
    }
    
    const code = this.props.conditions.weather[0].id
    if (code >= 700) {
      return background // normal
    } else if (code >= 600) {
      return backgroundSnow // snow
    } else if (code >= 200) {
      return backgroundRain // rain
    } else {
      return background
    }
  }

  render() {
    return(
      <section className="weatherToolbar" style={{backgroundImage: `url(${this.getImageSource()})`}}>
        <div className="container">
          <ZipCodeHistory
            zipCodeHistory={this.props.zipCodeHistory}
            zipEntered={this.props.zipEntered} />

          <form className="toolbar">
            <h1>Find Your Local Weather</h1>
            
            <div className="form-group">
              <label>Zip Code</label>
              <input
                type="text"
                placeholder="zip code"
                value={this.state.zipCode}
                onChange={this.zipChanged} />
              <button
                type="submit"
                disabled={this.zipLooksInvalid()}
                onClick={this.zipEntered}
                title="Get forecast for this location">Get Weather</button>
            </div>
          </form>
        </div>
      </section>
    )
  }

}
