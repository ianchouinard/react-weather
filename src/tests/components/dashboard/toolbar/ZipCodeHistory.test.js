import React from 'react';
import renderer from 'react-test-renderer';
import { ZipCodeHistory } from './../../../../components/dashboard/toolbar/ZipCodeHistory'
import { store } from '../../../../store/store'
import { setZipCode } from '../../../../actions/ForecastActions';

describe('ZipCodeHistory', () => {
  const component = renderer.create(
    <ZipCodeHistory />
  )

  it('should create the component', () => {
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  describe('zipSelected()', () => {
    it('should set the zip via the store', () => {
      jest.spyOn(store, 'dispatch')
      component.getInstance().zipSelected({preventDefault: () =>{}}, '03063')
      expect(store.dispatch).toHaveBeenCalledWith(setZipCode('03063'))
    })
  })

  describe('toggleMenu()', () => {
    it('should toggle the mobile menu state', () => {
      const currentState = component.getInstance().state.menuOpen
      component.getInstance().toggleMenu()

      expect(component.getInstance().state.menuOpen).toBe(!currentState)
    })
  })
})
