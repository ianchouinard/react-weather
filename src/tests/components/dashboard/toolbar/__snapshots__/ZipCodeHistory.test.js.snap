// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ZipCodeHistory should create the component 1`] = `
<nav
  className="zipCodeHistory"
>
  <button
    className="menuToggle "
    onClick={[Function]}
    title="Open Menu"
  >
    <span />
    <span />
    <span />
  </button>
  <span>
    Recent Locations
  </span>
  <ul
    className=""
  />
</nav>
`;
